***Admin functions for Biblosphere***

- MySQL functions 
-- Delete ISBN record 
-- Query book by ISBN

- API functions
-- Add book with ISBN
-- Recognize book shelf by photo
-- Scan not found ISBN

In [1]:
import firebase_admin
from firebase_admin import credentials
from firebase_admin import firestore
from firebase_admin import auth
from math import pi, sin, cos, atan2, sqrt
import requests
import re
import geohash2
from geopy.geocoders import Nominatim

import pandas as pd
import json


cred = credentials.Certificate("../biblosphere.secret/jupyter/firebase-adminsdk.json")
firebase_admin.initialize_app(cred)

class Book:
    def __init__(self, isbn, title, authors, image='', language=None):
        self.isbn = isbn
        self.title = title
        self.authors = authors
        self.image = image
        if language is not None:
            self.language = language
        self.keys = lexems(title + ' ' + authors + ' ' + isbn, full=True)

    @classmethod
    def from_json(cls, obj):
        return cls(obj['isbn'], obj['title'], obj['authors'], obj['image'])

    def catalog_title(self):
        return self.title + ' ' + self.authors
    
def lexems(s, full = False):
    if type(s) is str:
        return re.sub('[;()\",/&!?:.\-*·|+$\'«@•]',' ',s.lower()).split()
    elif type(s) is set:
        return [w.lower() for w in s]
    
    
# Function to calculate distance between two geo-points
def distance_between(lat1, lon1, lat2, lon2):
    r = 6378.137 # Radius of earth in KM
    d_lat = lat2 * pi / 180 - lat1 * pi / 180
    d_lon = lon2 * pi / 180 - lon1 * pi / 180
    a = sin(d_lat/2) * sin(d_lat/2) + cos(lat1*pi/180) * cos(lat2*pi/180) * sin(d_lon/2) * sin(d_lon/2)
    c = 2 * atan2(sqrt(a), sqrt(1 - a))
    distance = r * c

    return distance / 1000


def check_isbn_10(isbn):
    i, s, t = 0, 0, 0

    digits = [int(d) for d in isbn]
    
    for d in digits: 
        t += d
        s += t

    return s % 11


def calc_isbn_13(isbn):
    digits = [int(d) for d in isbn]
    
    check = 0
    for i, d in enumerate(digits[0:12]):
        #print(i, d)
        #print(d * (1 + 2 * (i % 2))
        check += d * (1 + 2 * (i % 2))
        
    return (10 - check % 10) % 10


# Test recognize:
# gcloud auth print-identity-token
headers = {"Authorization": "Bearer eyJhbGciOiJSUzI1NiIsImtpZCI6IjY1YjNmZWFhZDlkYjBmMzhiMWI0YWI5NDU1M2ZmMTdkZTRkZDRkNDkiLCJ0eXAiOiJKV1QifQ.eyJpc3MiOiJodHRwczovL2FjY291bnRzLmdvb2dsZS5jb20iLCJhenAiOiIzMjU1NTk0MDU1OS5hcHBzLmdvb2dsZXVzZXJjb250ZW50LmNvbSIsImF1ZCI6IjMyNTU1OTQwNTU5LmFwcHMuZ29vZ2xldXNlcmNvbnRlbnQuY29tIiwic3ViIjoiMTE2NDYwNDE2MTIxMzU4MTIyNjI1IiwiZW1haWwiOiJkc3RhcmsxOTc3QGdtYWlsLmNvbSIsImVtYWlsX3ZlcmlmaWVkIjp0cnVlLCJhdF9oYXNoIjoiSmFlNFFCYWNSRHdneEN4RS1ZeDZoZyIsImlhdCI6MTU5NDUyNTEwOSwiZXhwIjoxNTk0NTI4NzA5fQ.oby53OATR_4VfLZpiXPWJSac0PDG_MS6a5FBxIrdZYucy4-_5vVoggx1xNOOhOd7WRhrJ_6awYdf_eIes_kbVdPjfUsTslJUQCOc6tjH8sy69dqilwrSNqub1WPttcMUlq_t8qxKRlyZRL6a5uw7RjmU3MBJrjTyaNEq-XM1KrcVY1bW1Gwt7nAGGE8yqvuPzfVSyz-EgZ9fJFkl9hXDYhoBoL2mPv_43Z1dXBjK9QOcZ_3p4sKJKjERmXrWHANY-mA4__qeaTnpA-eh9nVr-H6CGFdOEmhRKNrHecW0QBi5aZYgLK3Af1DBDFUhaq5Ot34f4XZ81P7S1u--JmDtCA"}

def get_book(isbn):
    try:
        endpoint = "https://biblosphere-api-ihj6i2l2aq-uc.a.run.app/get?isbn=%s" % isbn

        res = requests.get(endpoint, headers=headers)

        if not res.ok:
            print('HTTP error:', res.status_code, res._content)
            return None
        
        res_json = json.loads(res._content)
        books = [Book.from_json(obj) for obj in res_json]

        if books is not None and len(books) > 0:
            return books[0]
        else:
            return None
    except Exception as e:
        print('Exception for book [%s]' % book['id'], e)
        traceback.print_exc()
        return None

    
def add_books(books):
    endpoint = "https://biblosphere-api-ihj6i2l2aq-uc.a.run.app/add"

    body = {'books': [
        {"isbn": b.isbn, 
         "title": b.title, 
         "authors": b.authors,
         "image": b.image,
        } for b in books]
    }
    res = requests.post(endpoint, headers=headers, json=body)
    
    if res.ok:
        print('%d books added' % len(books))
    else:
        print('HTTP error:', res.status_code, res._content)
        
    return    
    

def recognize_image(uid, image, shelf, point):
    endpoint = "https://biblosphere-api-ihj6i2l2aq-uc.a.run.app/add_user_books_from_image"
    body = {
              'uid': uid,
              'uri': 'images/%s/%s' % (uid, image),
              'shelf': shelf,
              'notification': False,
              'location': {
                  'lat': point.latitude,
                  'lon': point.longitude,
                  'geohash': geohash2.encode(point.latitude, point.longitude)[:9],
               }
            }
    
    #print('Input:', body)
    
    res = requests.post(endpoint, headers=headers, json=body)

    if res.ok:
        print(res._content)
        return True
    else:
        print('HTTP error:', res.status_code, res._content)
        return False

In [16]:
# Script to count recognized books for a period

db = firestore.client()

#books = db.collection('bookrecords').limit(30).stream()
books = db.collection('shelves').where('id', '>=', '1593006003543').stream()

count = 0
for i, b in enumerate(books):
    data = b.to_dict()
    if data['recognized'] is not None and data['recognized'] > 0:
        print('Recognized %d %s' % (data['recognized'], data['image']))
        count += data['recognized'] 

print('TOTAL RECOGNIZED:', count)

Recognized 8 images/IEbwfDjHQcQkTm4uGDMCQfMsGuY2/1593006003543.jpg
Recognized 7 images/IEbwfDjHQcQkTm4uGDMCQfMsGuY2/1593006083473.jpg
Recognized 6 images/IEbwfDjHQcQkTm4uGDMCQfMsGuY2/1593006110917.jpg
Recognized 5 images/IEbwfDjHQcQkTm4uGDMCQfMsGuY2/1593006125181.jpg
Recognized 13 images/IEbwfDjHQcQkTm4uGDMCQfMsGuY2/1593006137290.jpg
Recognized 17 images/IEbwfDjHQcQkTm4uGDMCQfMsGuY2/1593006145977.jpg
Recognized 1 images/IEbwfDjHQcQkTm4uGDMCQfMsGuY2/1593009273328.jpg
Recognized 3 images/IEbwfDjHQcQkTm4uGDMCQfMsGuY2/1593009413374.jpg
Recognized 4 images/IEbwfDjHQcQkTm4uGDMCQfMsGuY2/1593017157610.jpg
Recognized 10 images/IEbwfDjHQcQkTm4uGDMCQfMsGuY2/1593017363983.jpg
Recognized 5 images/IEbwfDjHQcQkTm4uGDMCQfMsGuY2/1593017975066.jpg
Recognized 1 images/IEbwfDjHQcQkTm4uGDMCQfMsGuY2/1593018196053.jpg
Recognized 2 images/IEbwfDjHQcQkTm4uGDMCQfMsGuY2/1593018204373.jpg
Recognized 1 images/IEbwfDjHQcQkTm4uGDMCQfMsGuY2/1593018213743.jpg
Recognized 7 images/ZN7Ux33RKWaaRT8ENjG0ToZvWzS2/1593278933

In [40]:
emails = ['webdoedie@gmail.com',
'camilleguzman.10880@gmail.com',
'angiewelch.88990@gmail.com',
'samueldixon.34540@gmail.com',
'skaplichniy@gmail.com',
'sof5859@yandex.ru',
'gingeralina@mail.ru',
'kanhai.kcc@gmail.com',
'anbarbery@gmail.com',
'mchusick@gmail.com',
'thomashibbard89@gmail.com',
'kolbasa12393@mail.ru',
'bdinh14@gmail.com',
'ashima.odsharma@gmail.com',
'mart.a.o.3393@gmail.com',
'poleshchukolala@gmail.com',
'pelageya99@gmail.com',
'topoeva.n@yandex.ru',
'kamiomisuzu@yandex.ru',
'avtarcute777@gmail.com',
'romanziouzev@gmail.com',
'd.tarusov@gmail.com',
'melaniedarklove@gmail.com',
'valentina.vosk@gmail.com',
'monkphx@gmail.com',
'mayisaru@gmail.com',
'zaxquit@gmail.com',
'nivadow@gmail.com',
'ranjansahabandu@gmail.com',
'taras8055@gmail.com',
'xodacevich@yandex.ru',
'danmlisa2@gmail.com',
'acar_4788@hotmail.com',
'jianing.qi@gridmachina.com',
'roydiego@gmail.com',
'julienhardy94@gmail.com',
'gaivoronskystan@gmail.com']

geolocator = Nominatim(user_agent="Biblosphere")

for e in emails:
    auth_user = auth.get_user_by_email(e)
    if auth_user is None:
        print('!!! User auth missed for', e)
        continue
    user = db.collection('users').document(auth_user.uid).get()
    if user is None:
        print(f'!!! User record missed for {auth_user.display_name} ({e})')
        continue
    
    data = user.to_dict()
    
    # Find boos and wishes
    records = db.collection('bookrecords').where('holderId', '==', auth_user.uid).stream()
    records = [b.to_dict() for b in records]

    books = [b for b in records if not b['wish']]
    wishes = [b for b in records if b['wish']]

    # Skip if no books/wishes added
    if len(records) == 0:
        continue

    print('-------------------------------------------------------------')
    print(f'{auth_user.display_name} ({e})')

    # Show location
    location = geolocator.reverse(f"{records[0]['location']['geopoint'].latitude}, {records[0]['location']['geopoint'].longitude}", language='en', addressdetails=True)
    #print(location.raw)
    country = location.raw['address']['country']

    if 'state' in location.raw['address']:
        state = location.raw['address']['state']
    elif 'state' in location.raw['address']:
        state = location.raw['address']['state_district']
    else:
        state = location.raw['address']['city']
    
    if 'city' in location.raw['address']:
        city = location.raw['address']['city']
    elif 'town' in location.raw['address']:
        city = location.raw['address']['town']
    elif 'suburb' in location.raw['address']:
        city = location.raw['address']['suburb']
    else:
        city = location.raw['address']['county']
    print(f'LOCATION: {city}, {state}, {country}')
    print(f'ADDRESS: {location.raw["address"]}')

    # Show books
    if len(books) > 0:
        print('BOOKS:')
        for b in books:
            print(f'"{b["title"]}" by {",".join(b["authors"])}')
            
    # Show wishes
    if len(wishes) > 0:
        print('WISHES:')
        for b in wishes:
            print(f'"{b["title"]}" by {",".join(b["authors"])}')

    # TODO: Show chats
    


-------------------------------------------------------------
Kanhai Kumar (kanhai.kcc@gmail.com)
LOCATION: Bengaluru, Karnataka, India
ADDRESS: {'neighbourhood': 'Vaikuntam Layout', 'suburb': 'Dodda Nekkundi', 'city_district': 'Mahadevapura Zone', 'city': 'Bengaluru', 'county': 'Bangalore East', 'state_district': 'Bangalore Urban', 'state': 'Karnataka', 'postcode': '560 037', 'country': 'India', 'country_code': 'in'}
BOOKS:
"Paradise Park" by Allegra Goodman
"The Power of Logical Thinking: Easy Lessons in the Art of Reasoning...and Hard Facts About Its Absence in Our Lives" by Marilyn Vos Savant
"Wild Designs: A Novel" by Katie Fforde
"Proud Mary" by Iris Gower
"Falling for You" by Jill Mansell
"English Is Easy" by Chetananand Singh
-------------------------------------------------------------
Thomas Hibbard (thomashibbard89@gmail.com)
LOCATION: Salta, Salta, Argentina
ADDRESS: {'house_number': '453', 'road': 'General José María Paz', 'city_district': 'B Calixto Gauna', 'city': 'Salta

-------------------------------------------------------------
Александра Мартынова (mart.a.o.3393@gmail.com)
LOCATION: Srednaya Rogatka, Saint Petersburg, Russia
ADDRESS: {'house_number': '27', 'road': 'улица Фрунзе', 'suburb': 'Srednaya Rogatka', 'municipality': 'округ Московская застава', 'state_district': 'Московский район', 'state': 'Saint Petersburg', 'region': 'Northwestern Federal District', 'postcode': '190000', 'country': 'Russia', 'country_code': 'ru'}
BOOKS:
"Дом, в котором..." by Mariam Petrosyan
"Усвятские шлемоносцы (сборник)" by Евгений Носов
-------------------------------------------------------------
Настя Топоева (topoeva.n@yandex.ru)
LOCATION: Заневское городское поселение, Leningrad oblast, Russia
ADDRESS: {'building': 'Дом у Каштановой аллеи', 'house_number': '1', 'road': 'Областная улица', 'city': 'Заневское городское поселение', 'county': 'Vsevolozhsky District', 'state': 'Leningrad oblast', 'region': 'Northwestern Federal District', 'postcode': '188689', 'count

-------------------------------------------------------------
Xodasevich Book Sharing (xodacevich@yandex.ru)
LOCATION: Хитровка, Moscow, Russia
ADDRESS: {'house_number': '6', 'road': 'Pokrovka Street', 'neighbourhood': 'Ivanovskaya Gorka', 'suburb': 'Хитровка', 'state_district': 'Central Administrative Okrug', 'state': 'Moscow', 'region': 'Central Federal District', 'postcode': '101990', 'country': 'Russia', 'country_code': 'ru'}
BOOKS:
"Бунт Афродиты. Tunk / Лоренс Даррелл ; " by Даррелл, Лоренс Джордж (1912-1990).
"Мы против вас" by Бакман Фредрик
"Больно только когда смеюсь" by Рубина Дина Ильинична
"Медвежий угол" by Фредрик Бакман
"Вероника решает умереть" by Коэльо Пауло
"Похороните меня за плинтусом" by Санаев Павел  Владимирович
"Земля имеет форму чемодана" by 
"Дело о чертовом зеркале" by 
"Госпожа Бовари" by Флобер Гюстав
"А зори здесь тихие... С реальными историями о женщинах на войне" by Васильев Борис Львович
"Инсайдер : роман / Стивен Фрей; " by Фрей,Стивен.
"В поисках Ал

In [2]:
# Admin script to collect all unique locations for the website

db = firestore.client()

markers = pd.DataFrame(columns=('lat', 'lng'))

#books = db.collection('bookrecords').limit(30).stream()
books = db.collection('bookrecords').stream()

for i, b in enumerate(books):
    data = b.to_dict()
    if data['location'] is not None:
        #print({'lat': data['location']['geopoint'].latitude, 'lng': data['location']['geopoint'].longitude})
        markers.loc[i] = [data['location']['geopoint'].latitude, data['location']['geopoint'].longitude]

print('Full length:', len(markers))
markers.drop_duplicates(inplace=True)
print('Without duplicates:', len(markers))

json = markers.to_json(orient='records')
#print(json)

with open('locations.json', 'w') as f:
    f.write(json)

Full length: 1948
Without duplicates: 623


In [2]:
# Admin script to get all books around the location

from math import pi, sin, cos, atan2, sqrt

# Function to calculate distance between two geo-points
def distance_between(lat1, lon1, lat2, lon2):
    r = 6378.137 # Radius of earth in KM
    d_lat = lat2 * pi / 180 - lat1 * pi / 180
    d_lon = lon2 * pi / 180 - lon1 * pi / 180
    a = sin(d_lat/2) * sin(d_lat/2) + cos(lat1*pi/180) * cos(lat2*pi/180) * sin(d_lon/2) * sin(d_lon/2)
    c = 2 * atan2(sqrt(a), sqrt(1 - a))
    distance = r * c

    return distance

#lat, lon, distance = 60.1699, 24.9384, 80.0 # Helsinki
lat, lon, distance = 25.2048, 55.2708, 80000.0 # Dubai


db = firestore.client()

#books = db.collection('bookrecords').limit(10).stream()
books = db.collection('bookrecords').stream()

count = 0
for i, b in enumerate(books):
    data = b.to_dict()
    if data['location'] is not None:
        d = distance_between(data['location']['geopoint'].latitude, data['location']['geopoint'].longitude, lat, lon)
        if d < distance and data['ownerId'] != 'SJIyyqAGPTcXWaVnha5pWe4kuIC3':
            #print('%.2f %s %s [%s]' % (d, data['authors'], data['title'], data['ownerId']))
            count += 1
        
print(count, 'books found')

2464 books found


In [ ]:
Robert T. Kiyosaki: Rich Dad, Poor Dad: What the Rich Teach Their Kids About Money--That the Poor and Middle Class Do Not!
Peter Thiel, Blake  Masters: Zero to One: Notes on Startups, or How to Build the Future

In [10]:
# Admin script to get all book exchanges and its context

def print_chat(u1, u2):
    chat_id = ':'.join(sorted([u2,u1]))
    
    messages = db.collection('messages').document(chat_id).collection(chat_id).order_by('timestamp').stream()
    for m in messages:
        data = m.to_dict()
        print(data['content'])

db = firestore.client()

books = db.collection('bookrecords').where('lent', '==', True).stream()

for i, b in enumerate(books):
    data = b.to_dict()
    # Exclude testing ids
    if data['ownerId'] == 'rzvlcuAdFVS6238qqse4B4EjvUA3' or data['ownerId'] == 'oyYUDByQGVdgP13T1nyArhyFkct1'  or \
        data['holderId'] == 'rzvlcuAdFVS6238qqse4B4EjvUA3' or data['holderId'] == 'oyYUDByQGVdgP13T1nyArhyFkct1':
        continue
        
    # Print book details
    print('***************************************************************************')
    print('BOOK: %s' % data['title'])
        
    # Print names and emails of the users
    owner = auth.get_user(data['ownerId'])
    if owner is not None:
        print('BOOK OWNER: %s (%s)' % (owner.display_name, owner.email))
        
    holder = auth.get_user(data['holderId'])
    if holder is not None:
        print('BOOK HOLDER: %s (%s)' % (holder.display_name, holder.email))
        
        
    print_chat(data['holderId'], data['ownerId'])
    
    # Print latest conversation of the users

***************************************************************************
BOOK: Письма к С. А. Толстой 1887-1910
BOOK OWNER: Марат Сабиров (avkhatovich@gmail.com)
BOOK HOLDER: Екатерина Баранова (baranova-1@bk.ru)
***************************************************************************
BOOK: Ð Ñ Ð¾ Ð¼Ñ Ñ Ð¾Ð½ÐºÐ° Рона, Ð Ñ Ð¾ Ð°Ð¼Ñ Ñ ÐµÑ Ð Ð°Ð¼Ñ ÐºÐ¾Ð Ð¾ ÐºÐ¾Ñ Ð° Тома, Ð Ð Ñ Ð¾ Ñ Ð°Ð Ð½Ð¾Ðµ Ð Ñ Ñ Ð Ð¾Ðµ (Ð ÐµÑ Ñ ÐºÐ Ðµ Ñ Ñ Ð Ñ Ð ) | Pro mishonka Shona, Pro amsterdamskogo kota Toma i Pro raznoe drugoe (children poetry) | Russian
BOOK OWNER: Анастасия Приказчикова (prikazav@gmail.com)
BOOK HOLDER: Алена Рудюк (alenar91@mail.ru)
***************************************************************************
BOOK: Ампир V. Vампир / Виктор Пелевин. - Москва : Эксмо, 2010. - 408 с.; 21 см. - (Новый Пелевин).; ISBN 978-5-699-40130-7
BOOK OWNER: Людмила Дадова (dadovalyuda@rambler.ru)
BOOK HOLDER: Вероника Эссен (None)
******************************************************************

In [14]:
# Admin script to check messages

db = firestore.client()

#chats = db.collection('messages').where('toId', '==', 'xrTVCEyCwfOtO8lDe5UKk2egF4c2').stream()
chats = db.collection('messages').stream()


for c in chats:
    data = c.to_dict()
    u1, u2 = data['fromId'], data['toId']

    # Skip incomplete records
    if u1 is None or u2 is None:
        #print('toId or fromId is missing')
        continue
    
    # Skip test users
    if u1 in ['rzvlcuAdFVS6238qqse4B4EjvUA3', 'TzbOEGICy0XVPCUA6XUbTOKPXap2', 'oyYUDByQGVdgP13T1nyArhyFkct1', '0000000000000000000000000000'] or \
        u2 in ['rzvlcuAdFVS6238qqse4B4EjvUA3', 'TzbOEGICy0XVPCUA6XUbTOKPXap2', 'oyYUDByQGVdgP13T1nyArhyFkct1', '0000000000000000000000000000']:
        continue
    
    message_ref = c.reference.collection(c.id)
        
    if message_ref is not None:
        results = message_ref.order_by('timestamp').stream()
        messages = [m.to_dict() for m in results]
        
        if len(messages) > 0:
            user1 = auth.get_user(u1)
            if user1 is None:
                #print('Auth user for %s NOT found' % u1)
                continue

            user2 = auth.get_user(u2)
            if user2 is None:
                #print('Auth user for %s NOT found' % u2)
                continue

            print('************************************************************')
            print('CHAT BETWEEN: %s (%s) and %s (%s)' % (user1.display_name, user1.email, user2.display_name, user2.email))

            for m in messages:
                print(m['content'])


************************************************************
CHAT BETWEEN: Владислав Макулов (vmagent80@gmail.com) and Айрат Денисламов (laryartufa@gmail.com)
Привет!
Рад, что ты здесь!
Надо пополнять сообщество)
привет! 
мне очень понравилась идея! 
друзьям теперь показываю и рассказываю
нужно ещё книжки добавить
как у тебя дела? 
похоже в этом приложении нифига нет оповещений
дела хорошо, тоже думаю побольше книг выложить
да, нет оповещений
************************************************************
CHAT BETWEEN: Alfiya Kohanidze (ar.valiullina@gmail.com) and Владислав Макулов (vmagent80@gmail.com)
Альфия, привет!
Эта книга как-то связана с "Магической уборкой" ля взрослых?
Марии Кондо?
************************************************************
CHAT BETWEEN: Xodasevich Book Sharing (xodacevich@yandex.ru) and Стасио Гайворонски (gaivoronskystan@gmail.com)
Стас привет проверяю приложение 
есс
************************************************************
CHAT BETWEEN: Olga Pylaeva (ol

************************************************************
CHAT BETWEEN: Ирина Реброва (irinare86@gmail.com) and Anastasia Kostsova (nastya-kostsova@yandex.ru)
Можно взять у вас "Семейный тайм-менеджмент. Книга для родителей, которые хотят «все успеть»"?
************************************************************
CHAT BETWEEN: Vadim Mikhailov (dingoinlimbo@gmail.com) and Людмила Дадова (dadovalyuda@rambler.ru)
Добрый день! у вас очень интересные книги ) я Вадим. и наверняка вы хорошо знаете Дениса с Женей)
Привет!
Да, знаю.но в основном Женю)
************************************************************
CHAT BETWEEN: Margarita Ozhima (ozhimamargarita@gmail.com) and Хорошая Жена (goodwifeme@gmail.com)
привет ) могу дать почитать
************************************************************
CHAT BETWEEN: Ekaterina Gureeva (irracio.n@gmail.com) and Anastasia Gusakova (anastasiadatsik@gmail.com)
здравствуйте, у вас такая тематическая полка чудесная! 
как вам «после трёх уже поздно»?
здрав

In [3]:
# Admin script to go through not identified ISBN and print it

db = firestore.client()

#books = db.collection('bookrecords').limit(30).stream()
isbns = db.collection('noisbn').where('found', '==', True).stream()

isbns = [b for b in isbns]

print('Number of records:', len(isbns))

count, resolved = 0, 0

# List of users with books to add
# {uid: {name: <name>, email: <email>, books: []}}
users = {}

for b in isbns:
    isbn = b.id
    data = b.to_dict()
    
    if len(isbn) == 10 and check_isbn_10(isbn) == 0:
        isbn = '978' + isbn
        isbn = isbn[:12] + str(calc_isbn_13(isbn))
        print('%s => %s' % (b.id, isbn))

    if len(isbn) != 13 or not isbn.startswith('97'):
        #print('ISBN %s are incorrect' % isbn)
        continue

    if 'requested_by' not in data or 'found' not in data or not data['found']:
        #print('ISBN %s has no users' % isbn)
        continue
        
    book = get_book(isbn)
    if book is not None:
        #print('%s requested by' % isbn, data['requested_by'])
        for uid in data['requested_by']:
            user = auth.get_user(uid)
            if user is not None:
            #print('%s requested by (%s, %s, %s)' % (isbn, uid, user.display_name, user.email))
                if uid in users:
                    users[uid]['books'].append(book)
                else:
                    users[uid] = {'name': user.display_name, 'email': user.email, 'books': [book]}
            else:
                print('Missing user:', uid)
    else:
        print('Book missing:', isbn)

Number of records: 86
HTTP error: 504 b'upstream request timeout'
Book missing: 9785222234976


In [13]:
i = 0
for uid in users:
    # Skip some users
    if uid in ['ot0jztXgSrMRYIWs3QwIRwTvimN2', 'mRGmf2rI7xSGqDq95NL5RTK43jl2', 'GO8apnl0KcU0uwfOyJWORUHAUGG3',
              'qKKPX23jaWO2VqYJShWKouup2P32']:
        continue
        
    data = users[uid]
    #print('%s has %d books to add' % (data['name'], len(data['books'])))
    
    user = db.collection('users').document(uid).get()
    if not user.exists:
        print(f'User record missing: {uid}, {data["name"]}')
    else:
        user_data = user.to_dict()
        if 'position' in user_data and user_data['position'] is not None:
            users[uid]['position'] = user_data['position']
        else:
            users[uid]['position'] = None
            other_books = db.collection('bookrecords').where('holderId', '==', uid).limit(1).stream()
            other_books = [b for b in other_books]
            
            if len(other_books) > 0:
                other_data = other_books[0].to_dict()
                if 'location' in other_data and other_data['location'] is not None:
                    print(f'Geo-position found {uid}, {data["name"]}')
                    users[uid]['position'] = other_data['location']['geopoint']

            if users[uid]['position'] is None:
                print(f'Position missed for {uid}, {data["name"]}')

        if 'photo' in user_data:
            users[uid]['photo'] = user_data['photo']
        elif 'photoUrl' in user_data:
            users[uid]['photo'] = user_data['photoUrl']
        else:
            users[uid]['photo'] = ''
            print(f'Photo missed for {uid}, {data["name"]}')
    
    print('-------------------------------------------------------')
    print(data['email'])

    print(f"Hello {data['name']},\n\nYou've tryed to add books to Bibloshere app and some books were not found by ISBN. I've corrected problems with ISBN search and added below books into your catalog. Sorry for inconvenience. Let me know if I can help you with the app.\n")

    for i, b in enumerate(data['books']):
        print(f'{i+1}) {b.authors} "{b.title}"')
        
        b.keys.append(isbn)
        
        # Add bookrecord to Firestore
        rec = {
            'authors': [s.strip() for s in b.authors.split(',')],
            'confirmed': False,
            'holderId': uid,
            'holderName': data['name'],
            'holderImage': users[uid]['photo'],
            'id': f'b:{uid}:{isbn}',
            'image': b.image,
            'isbn': isbn,
            'keys': b.keys,
            'lent': False,
            'location': {
                  'geopoint': users[uid]['position'],
                  'geohash': geohash2.encode(users[uid]['position'].latitude, users[uid]['position'].longitude)[:9],
               },
            'matched': False,
            'matcheId': None,
            'ownerId': uid,
            'ownerName': data['name'],
            'ownerImage': users[uid]['photo'],
            'title': b.title,
            'transit': False,
            'users': [uid],
            'wish': False
        }
        db.collection('bookrecords').document(rec['id']).set(rec, merge=True)
        #print(f"Bookrecord added {rec['id']}, rec['authors']") 
        #print(f"AUTHORS {rec['authors']}") 

    print('\n- Denis Stark')

    #i += 1
    #if i >= 1:
    #    break
    #9785354014804
    


Geo-position found M8eWIMNYZrfelOEUg4T1EOusdxE3, Thomas Hibbard
-------------------------------------------------------
thomashibbard89@gmail.com
Hello Thomas Hibbard,

You've tryed to add books to Bibloshere app and some books were not found by ISBN. I've corrected problems with ISBN search and added below books into your catalog. Sorry for inconvenience. Let me know if I can help you with the app.

1) Hal Whitehead;Luke Rendell "The Cultural Lives of Whales and Dolphins"

- Denis Stark
-------------------------------------------------------
dstark1977@gmail.com
Hello Денис Старк,

You've tryed to add books to Bibloshere app and some books were not found by ISBN. I've corrected problems with ISBN search and added below books into your catalog. Sorry for inconvenience. Let me know if I can help you with the app.

1) Don Tapscott, Anthony D. Williams "Wikinomics: How Mass Collaboration Changes Everything"
2) Mohammed Bin Rashid Al Maktoum "Reflections on Happiness and Positivity"
3) Nat


- Denis Stark
-------------------------------------------------------
sergeevapd@gmail.com
Hello Полина Сергеева,

You've tryed to add books to Bibloshere app and some books were not found by ISBN. I've corrected problems with ISBN search and added below books into your catalog. Sorry for inconvenience. Let me know if I can help you with the app.

1) Денис Грушевский "Седьмое измерение"
2) Мариеа Миллер "100 Способов Очистить Дом От Энергетической Грязи"
3) Дрибноход, Юлия Юрьевна. "Искусство исцеления кожи : Пол. энцикл. : Траволечение. Водолечение. Голодание. Очищение. Глинолечение. Питание / Юлия Дрибноход. - СПб. : ВЕСЬ, 2002. - 255 с. : ил.; 26 см. - (Лечение без лекарств).; ISBN 5-266-00095-3"
4) Мураками, Харуки. "Охота на овец : роман / Мураками Х. - СПб. : Амфора, 2006 (СПб. : ГИПК Лениздат). - 380 с. - (Амфора 2006).; ISBN 5-367-00217-X (В пер.)"
5) Дайер Алан "Звёзды"
6) Juan-Eduardo Cirlot "Gaudi, vvodenie v ego arkitekturu"

- Denis Stark
Geo-position found IEbwfDjHQcQkTm

In [53]:
# Admin script to go through not identified ISBN and try to find it

db = firestore.client()

#isbns = db.collection('noisbn').limit(20).stream()
isbns = db.collection('noisbn').stream()

isbns = [b for b in isbns]

print('Number of records:', len(isbns))

count, resolved = 0, 0

to_delete = []
found = []
for b in isbns:
    isbn = b.id
    data = b.to_dict()

    # Skip records if it was sorted out before
    if 'found' in data:
        continue

    # Skip record if no users recorded for it
    if 'requested_by' not in data:
        continue

    if len(isbn) == 10 and check_isbn_10(isbn) == 0:
        isbn = '978' + isbn
        isbn = isbn[:12] + str(calc_isbn_13(isbn))
        print('%s => %s' % (b.id, isbn))

    if len(isbn) == 13 and isbn.startswith('97'):
        #To just print ISBN
        #print(isbn)
        #continue
        
        count += 1
        #print('%s try to resolve...' % isbn)
        book = get_book(isbn)

        if book is not None:
            resolved += 1
            found.append(b.id)
            print('ISBN %s: resolved' % isbn)
            db.collection('noisbn').document(isbn).update({'found': True})
        else:
            print('ISBN %s: NOT resolved' % isbn)
            db.collection('noisbn').document(isbn).update({'found': False})
    else:
        to_delete.append(b.id)
        

print('%d out of %d resolved' % (resolved, count))

Number of records: 195
ISBN 9781897510247: resolved
ISBN 9785001311379: resolved
ISBN 9785030038322: resolved
ISBN 9785170026418: resolved
ISBN 9785170511518: resolved
ISBN 9785222025673: NOT resolved
ISBN 9785222037355: resolved
ISBN 9785271451942: resolved
ISBN 9785389015234: resolved
ISBN 9785699563890: resolved
ISBN 9785887111773: NOT resolved
ISBN 9785905906572: NOT resolved
ISBN 9785906258205: NOT resolved
ISBN 9785906264527: resolved
ISBN 9785911031541: resolved
ISBN 9785917430621: resolved
ISBN 9785917592107: resolved
ISBN 9785943554575: resolved
ISBN 9785950054150: NOT resolved
ISBN 9785953347686: resolved
ISBN 9785979200064: resolved
ISBN 9785982120182: resolved
ISBN 9789088830310: NOT resolved
ISBN 9789665215622: resolved
ISBN 9789851301351: NOT resolved
ISBN 9789851522350: resolved
ISBN 9789851531659: resolved
ISBN 9791587120366: NOT resolved
ISBN 9795367002415: resolved
ISBN 9799856307197: NOT resolved
21 out of 30 resolved


In [52]:
# ADD BOOKS IN MYSQL (Manually found in web)

books = []
books.append(Book('9781897510247', 'Дао-Дэ-Цзин', 'Владимир Антонов', image='https://www.ellibs.com/sites/default/files/imagecache/product/bookcover_9781897510247.jpg', language='rus'))
books.append(Book('9785001311379', '21 Урок Для Xxi Века', 'Юваль Ной Харари', image='https://www.rahvaraamat.ee/images/products/001/354/892/thumbnails/big/0ab9d8016fd041b7e31ef2a77cddcf646baa2706/21-%D1%83%D1%80%D0%BE%D0%BA-%D0%B4%D0%BB%D1%8F-xxi-%D0%B2%D0%B5%D0%BA%D0%B0.jpg', language='rus'))
books.append(Book('9785030038322', 'Физиология Человека', '', image='https://images-na.ssl-images-amazon.com/images/I/415MWsuJk-L._SX298_BO1,204,203,200_.jpg', language='rus'))
books.append(Book('9785170026418', 'Ночной Портье', 'Ирвин Шоу', image='https://images-na.ssl-images-amazon.com/images/I/51kqNUUKtFL._SX301_BO1,204,203,200_.jpg', language='rus'))
books.append(Book('9785170511518', '100 Способов Очистить Дом От Энергетической Грязи', 'Мариеа Миллер', image='https://images-na.ssl-images-amazon.com/images/I/41JA9htKUvL._BO1,204,203,200_.jpg', language='rus'))
books.append(Book('9785222037355', 'Техника Быстрого Чтения', 'Андреев', image='', language='rus'))
books.append(Book('9785271451942', 'Ружья, Микробы И Сталь', 'Джаред Даймонд', image='https://www.rahvaraamat.ee/images/products/000/067/415/thumbnails/view/ae1f30730f4669ffb44de832d34af385aa4475df/%D1%80%D1%83%D0%B6%D1%8C%D1%8F-%D0%BC%D0%B8%D0%BA%D1%80%D0%BE%D0%B1%D1%8B-%D0%B8-%D1%81%D1%82%D0%B0%D0%BB%D1%8C.jpg', language='rus'))
books.append(Book('9785389015234', 'Звёзды', 'Дайер Алан', image='https://images-na.ssl-images-amazon.com/images/I/3156Utn8GRL._BO1,204,203,200_.jpg', language='rus'))
books.append(Book('9785699563890', 'Искусство Грудного Вскармливания', 'Тереза Питман, Диана Вест, Дайен Виссингер', image='https://www.rahvaraamat.ee/images/products/000/084/180/thumbnails/view/ce340ce3290c5b1909fc829894c1b6dbc7963bb0/%D0%B8%D1%81%D0%BA%D1%83%D1%81%D1%81%D1%82%D0%B2%D0%BE-%D0%B3%D1%80%D1%83%D0%B4%D0%BD%D0%BE%D0%B3%D0%BE-%D0%B2%D1%81%D0%BA%D0%B0%D1%80%D0%BC%D0%BB%D0%B8%D0%B2%D0%B0%D0%BD%D0%B8%D1%8F.jpg', language='rus'))
books.append(Book('9785906264527', 'Голодный Город. Как Еда Определяет Нашу Жизнь', 'Кэролин Стил', image='https://pictures.abebooks.com/isbn/9785906264527-us.jpg', language='rus'))
books.append(Book('9785911031541', 'Культура Маркетинга. Маркетинг Культуры', 'Джон Сибрук', image='https://www.rahvaraamat.ee/images/products/000/697/105/thumbnails/view/9408fabc13d5dd5720304b128bbca92fe2001b2a/%D0%BA%D1%83%D0%BB%D1%8C%D1%82%D1%83%D1%80%D0%B0-%D0%BC%D0%B0%D1%80%D0%BA%D0%B5%D1%82%D0%B8%D0%BD%D0%B3%D0%B0-%D0%BC%D0%B0%D1%80%D0%BA%D0%B5%D1%82%D0%B8%D0%BD%D0%B3-%D0%BA%D1%83%D0%BB%D1%8C%D1%82%D1%83%D1%80%D1%8B-nobrow.jpg', language='rus'))
books.append(Book('9785917430621', 'На Стороне Подростка', 'Франсуаза Дольто', image='https://libs.ru/book/441/cover_441202.jpg', language='rus'))
books.append(Book('9785917592107', 'Мой Внутренний Элвис', 'Яна Шерер', image='https://www.knigi-club.ru/thumb/324x477xCUT/upload/iblock/8a7/moy_vnutrenniy_elvis_.jpeg', language='rus'))
books.append(Book('9785943554575', 'Брак Умер... Да Здравствует Семья!', 'Анатолий Некрасов', image='https://www.knygy.com.ua/pix/0c/bf/c9/0cbfc9ea4445d3a199f77c47e7c0a5ab.jpg', language='rus'))
books.append(Book('9785953347686', '100 Великих Людеи', 'Сергей Мусский', image='https://images-na.ssl-images-amazon.com/images/I/51MDNUk-0BL._SX298_BO1,204,203,200_.jpg', language='rus'))
books.append(Book('9785979200064', 'Издание Журнала. От Идеи До Воплощения', 'Джон Морриш', image='https://libs.ru/book/731/cover_731092.jpg', language='rus'))
books.append(Book('9785982120182', 'Любовныи Многоуголник', 'Анатолий Некрасов', image='https://images-na.ssl-images-amazon.com/images/I/31E+8MqNSLL._BO1,204,203,200_.jpg', language='rus'))
books.append(Book('9789665215622', 'Третье Открытие Силы', 'Андрей Сидерский', image='https://images-na.ssl-images-amazon.com/images/I/41nNk9j8A6L._BO1,204,203,200_.jpg', language='rus'))
books.append(Book('9789851522350', 'Супермышление', 'Тони Бьюзен, Барри Бьюзен', image='https://images-na.ssl-images-amazon.com/images/I/515QIA5P7mL._SX335_BO1,204,203,200_.jpg', language='rus'))
books.append(Book('9789851531659', '7 Навыков Высокоэффективных Семей', 'Стивен Р Кови ', image='https://www.troykaonline.com/files/product/large/246789_9789851531659.jpg', language='rus'))
books.append(Book('9795367002415', 'Русские Летописи Xi-Xvi Веков', 'А. Боброва', image='https://images-na.ssl-images-amazon.com/images/I/61ynENagZXL._SX339_BO1,204,203,200_.jpg', language='rus'))

add_books(books)

#book = get_book('9781897510247')
#if book is not None:
#    print(book.title, book.authors)


21 books added


181 out of 63 resolved
63 12
ISBN 9780226325927 Requested by ['M8eWIMNYZrfelOEUg4T1EOusdxE3']
ISBN 9780812928082 Requested by ['ot0jztXgSrMRYIWs3QwIRwTvimN2']
ISBN 9781119209591 Requested by ['ot0jztXgSrMRYIWs3QwIRwTvimN2']
ISBN 9781565921511 Requested by ['mRGmf2rI7xSGqDq95NL5RTK43jl2']
ISBN 9782745976192 Requested by ['oyYUDByQGVdgP13T1nyArhyFkct1']
ISBN 9783426274521 Requested by ['8UJw0yj5ceVq2lDq5KQfLlB5UE52']
ISBN 9785000572054 Requested by ['vDWiB1lV2yYAAd8t7J3PkeKhRpH2']
ISBN 9785001006091 Requested by ['xem1khYLX0a3d7KuE2FTYa9QGcK2']
ISBN 9785090117647 Requested by ['noaN3PE4YRQcOa4dtTeaPl5d2zN2']
ISBN 9785170164233 Requested by ['8UJw0yj5ceVq2lDq5KQfLlB5UE52']
ISBN 9785170259786 Requested by ['IEbwfDjHQcQkTm4uGDMCQfMsGuY2']
ISBN 9785170804658 Requested by ['vDWiB1lV2yYAAd8t7J3PkeKhRpH2']
ISBN 9785170896370 Requested by ['vDWiB1lV2yYAAd8t7J3PkeKhRpH2']
ISBN 9785171164997 Requested by ['8UJw0yj5ceVq2lDq5KQfLlB5UE52']
ISBN 9785222136591 Requested by ['noaN3PE4YRQcOa4dtTeaPl5d2zN

In [62]:
# Admin script to add all books from Firestore to MySQL

db = firestore.client()

#books = db.collection('bookrecords').limit(30).stream()
books = db.collection('books').stream()
books = [b for b in books]

print('Number of records:', len(books))

books_to_add = []

for b in books:
    data = b.to_dict()
    
    if 'migrated' in data and data['migrated']:
        continue
     
    if 'isbn' in data and ('title' in data or 'authors' in data):
        #print('Authors:', data['authors'])
        if 'authors' in data and data['authors'] is not None:
            authors = ','.join([a for a in data['authors'] if a is not None])
        else:
            authors = ''
        books_to_add.append(Book(data['isbn'], data['title'], authors, data['image']))

        if 'authors' in data and len(data['authors']) > 1:
            print('Multi-authors: %s' % data['isbn'], data['authors'])
    elif 'book' in data:
        data = data['book']
        if 'isbn' in data and ('title' in data or 'authors' in data):
            if 'authors' in data and data['authors'] is not None:
                authors = ','.join([a for a in data['authors'] if a is not None])
            else:
                authors = ''
            books_to_add.append(Book(data['isbn'], data['title'], authors, data['image']))

            if 'authors' in data and len(data['authors']) > 1:
                print('Multi-authors: %s' % data['isbn'], data['authors'])
    else:
        print('DATA MISSING:', data)
    
    if len(books_to_add) >= 10:
        add_books(books_to_add)
        for isbn in [b.isbn for b in books_to_add]:
            db.collection('books').document(isbn).update({'migrated': True})

        books_to_add = []

if len(books_to_add) > 0:        
    add_books(books_to_add)
    for isbn in [b.isbn for b in books_to_add]:
        db.collection('books').document(isbn).update({'migrated': True})


Number of records: 909


In [98]:
# Admin script to run shelf recognition for old shelves

db = firestore.client()


shelves = db.collection('shelves').where('user', '>', '').where('status', '==', 5).stream()
#shelves = db.collection('shelves').where('user', '>', '').stream()
shelves = [s for s in shelves]

print('Number of book shelves:', len(shelves))

for s in shelves:
    data = s.to_dict()
    
    # Skip shelves already requested for processing
    #if 'status' in data:
    #    continue

    # Run recognition via API
    print('Run recognition for user [%s], image [%s], shelf [%s], at coordinates [%.5f, %.5f]' % (data['user'], data['file'], s.id, data['position'].latitude, data['position'].longitude))
    res = recognize_image(data['user'], data['file'], s.id, data['position'])

Number of book shelves: 2
Run recognition for user [wuwImhUnaNVYCV0eDy6ydoGLqYl1], image [1547731544738.jpg], shelf [-LWQzJKadlza372rDoPr], at coordinates [55.82460, 37.50127]
b'Image recognition requested'
Run recognition for user [zKhrMV3lusY7QyMqnO1kCHV8hs52], image [1555069784600.jpg], shelf [-LcGNSxjnTCPL-S15bGE], at coordinates [60.07477, 30.33278]
b'Image recognition requested'


In [3]:
# Run recognition for one book shelf
#recognize_image('NW3sjiR8NQT6OwZyRFwCEmgbZ8j1', '1589726917860.jpg', '1589726917860', firestore.GeoPoint(43.7113831, 20.6709329))
recognize_image('0000000000000000000000000000', '0000000000034.jpg', '0000000000000000000000000000:0000000000034', firestore.GeoPoint(90.0, 135.0))

# Add books
#books = [Book('9785946230469', 'Энциклопедия для детей. Т. 23: Универсальный иллюстрированный энциклопедический словарь', '', image='https://images-na.ssl-images-amazon.com/images/I/21HqXNQA5FL._BO1,204,203,200_.jpg')]
#add_books(books)

# Get book by ISBN
#book = get_book('9785946230469')
#print(book.isbn, book.title)

b'Image recognition requested'


True

In [99]:
# Script to check results of the recognition

db = firestore.client()

shelves = db.collection('shelves').where('user', '>', '').stream()
shelves = [s for s in shelves]

print('Number of book shelves:', len(shelves))

recognized, total, completed, failed, lookup = 0, 0, 0, 0, 0
for s in shelves:
    data = s.to_dict()
    
    if data['status'] == 6:
        completed += 1
        recognized += data['recognized']
        total += data['total']
    elif data['status'] == 7:
        failed += 1
    else:
        print('Other status:', data['status'])

print('Completed: %d, Failed: %d, Lookup: %d' % (completed, failed, lookup))
print('Recognized: %d (%d%%), Total: %d'  % (recognized, 100 * recognized / total, total))



Number of book shelves: 170
Completed: 170, Failed: 0, Lookup: 0
Recognized: 1010 (50%), Total: 2017


In [4]:
# Admin script to visualize recognition
# Get the image and recognition json from GCS and visualize
import cv2
import numpy as np
from firebase_admin import storage

def imread_blob(blob):
    img = cv2.imdecode(np.asarray(bytearray(blob.download_as_string()), dtype=np.uint8), cv2.IMREAD_COLOR)
    return img

bucket = storage.bucket('biblosphere-210106.appspot.com')

#user = '0000000000000000000000000000'
#shelf = '0000000000034'

user = 'AWsv5n9QiJYIUcDQmxWlBKRoBzZ2'
shelf = '1590618924998' # '1590619142574', '1590619327951'
#shelf = '1590619142574' # '1590619327951'
#shelf = '1590619327951'

image_blob = bucket.blob('images/%s/%s.jpg' % (user, shelf))
result_blob = bucket.blob('images/%s/%s.json' % (user, shelf))

img = imread_blob(image_blob)
results = json.loads(result_blob.download_as_string())

for b in results['recognized']:
    cv2.drawContours(img, np.array([b['outline']]), 0, (0, 255, 0), 7)

for b in results['unrecognized']:
    cv2.drawContours(img, np.array([b['outline']]), 0, (0, 0, 255), 7)

cv2.imwrite('results/'+user+'-'+shelf+'.jpg', img)



True

In [17]:
for b in results['recognized']:
    print(b['book']['isbn'], '"%s"' % b['book']['title'])

9781565922204 "Advanced Perl Programming"
9781558605824 "GUI Bloopers: Don'ts and Do's for Software Developers and Web Designers"
9781449342685 "Microinteractions: Designing with Details"
9780963124685 "Those Who Trespass: A Novel of Murder and Television"
9780596000356 "Information Architecture for the World Wide Web: Designing Large-Scale Web Sites"
9780735710368 "Photoshop 6 Web Magic (Magic (New Riders))"
9781510710306 "My Robot Ate My Homework: Project Droid #3 (Paperback)"
9781565922570 "Mastering Regular Expressions"
9780596551872 "Apache Cookbook"
9781118026687 "Professional jQuery"
9781491913550 "Information Architecture"
9781401322915 "Free: The Future of a Radical Price"
9780201657968 "XML and SQL: Developing Web Applications"
9781731036841 "Introduction to Data Science with Python: Basics of Numpy and Pandas (Paperback)"
9780596004477 "Google Hacks: 100 Industrial-Strength Tips &amp; Tools"
9780470843710 "Web Programming: Building Internet Applications, 2nd Editon"
97815659

**FIRST RUN (Error in web-search)**
* Number of book shelves: 171
* Completed: 117, Failed: 51
* Recognized: 560 (40%), Total: 1384

**SECOND RUN (only processed schelves have been taken)**
* Number of book shelves: 170
* Completed: 170, Failed: 0, Lookup: 0
* Recognized: 1010 (50%), Total: 2017





In [ ]:
# TODO actions:

# 1. Corrupted symbols (abebooks):
# 9789874024770, 9785733103037

Potsdamer Platz, Bellevuestraße, Botschaftsviertel, Tiergarten, Mitte, Berlin, 10785, Deutschland
